### PR TITLE
pythonPackages.flatbuffers: init at 1.12.0

### PR DESCRIPTION
**Pull request by Wulfsta**
###### Motivation for this change

Expose [flatbuffers for python](https://pypi.org/project/flatbuffers/). It appears future TensorFlow builds will need this.
@teh 
@acowley 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by djanatyn** [comment]
I was able to successfully load this package, LGTM:
```
❯ nix repl
Welcome to Nix version 2.3.8. Type :? for help.

nix-repl> :l
Added 12691 variables.

nix-repl> :b pkgs.python37Packages.flatbuffers
[1 built, 13 copied (70.0 MiB), 12.8 MiB DL]

this derivation produced the following outputs:
  out -> /nix/store/9fmi2pigmakf0dnd2jq2lr3c0zscyiz5-python3.7-flatbuffers-1.12.0

nix-repl> :u pkgs.python38.withPackages(ps: with ps; [ python38Packages.flatbuffers ])
these derivations will be built:
  /nix/store/41wwylbsz2wgnkq1mz3yyjcbprmh7jy5-python3.8-flatbuffers-1.12.0.drv
  /nix/store/kip938br5c83scb34kf8hywh8zr25xyb-python3-3.8.6-env.drv
building '/nix/store/41wwylbsz2wgnkq1mz3yyjcbprmh7jy5-python3.8-flatbuffers-1.12.0.drv'...
...
OK
Finished executing setuptoolsCheckPhase
building '/nix/store/kip938br5c83scb34kf8hywh8zr25xyb-python3-3.8.6-env.drv'...
created 217 symlinks in user environment

[nix-shell:~/repos/nixpkgs]$ python
Python 3.8.6 (default, Sep 23 2020, 13:54:27)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import flatbuffers
>>> dir(flatbuffers)
['Builder', 'Table', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'builder', 'compat', 'compat_range', 'encode', 'number_types', 'packer', 'table', 'util']
>>> exit()
```

I'm curious if it's possible to pass in `flatbuffers` (instead of `pkgs`) without running into infinite recursion issues referencing it? I spent some time exploring what that might look like, but never got a successful build.

**Comment by Wulfsta** [comment]
@djanatyn The only thing I can think of that might be cleaner is a 
```
let 
  fltbuf = pkgs.flatbuffers;
in
...
```
but that doesn't change that `pkgs` has to be passed rather than `flatbuffers`.

**Comment by teh** [comment]
I think the naming convention is still `python3Packages.flatbuffers: init at 1.12.0` even if the source is derived like this. Otherwise LGTM.

**Comment by Wulfsta** [comment]
Changed commit message and improved the derivation.

File: pkgs/development/python-modules/flatbuffers/default.nix [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{ stdenv
+, buildPythonPackage
+, pkgs
+}:
+
+let
+  flatbuf = pkgs.flatbuffers;
```
Comment (IvarWithoutBones) [inline pkgs/development/python-modules/flatbuffers/default.nix:7]: You shouldn't be passing `pkgs` in a derivation like this.
```suggestion
, flatbuffers
}:

let
  flatbuf = flatbuffers;
```

Tested on my NixOS machine, works fine.

File: pkgs/development/python-modules/flatbuffers/default.nix [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+in buildPythonPackage rec {
+  pname = "flatbuffers";
+  version = flatbuf.version;
+  src = flatbuf.src;
+  sourceRoot = "source/python";
```
Comment (IvarWithoutBones) [inline pkgs/development/python-modules/flatbuffers/default.nix:12]: Some spacing would be nice.
```suggestion
  pname = "flatbuffers";
  version = flatbuf.version;
  
  src = flatbuf.src;
  sourceRoot = "source/python";
```

File: pkgs/development/python-modules/flatbuffers/default.nix [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (IvarWithoutBones) [inline pkgs/development/python-modules/flatbuffers/default.nix:13]: No unit tests seem to be available, you should use `pythonImportsCheck`.
```suggestion
  pythonImportsCheck = [ "flatbuffers" ];
```

**Review by IvarWithoutBones — CHANGES_REQUESTED** [review]
A few nits :)

File: pkgs/development/python-modules/flatbuffers/default.nix (hunk 1 above)
Comment (Wulfsta) [reply pkgs/development/python-modules/flatbuffers/default.nix:7]: This yields an infinite recursion on my machine. I'm calling this with the following:
```
nix-shell -I nixpkgs=../nixpkgs -p 'python3.withPackages(ps: with ps; [ flatbuffers ])' --show-trace
```

Can you further investigate or explain why I would see an infinite recursion here?

File: pkgs/development/python-modules/flatbuffers/default.nix (hunk 2 above)
Comment (Wulfsta) [reply pkgs/development/python-modules/flatbuffers/default.nix:12]: Will fix once the top issue is investigated.

File: pkgs/development/python-modules/flatbuffers/default.nix (hunk 3 above)
Comment (Wulfsta) [reply pkgs/development/python-modules/flatbuffers/default.nix:13]: Will fix once the top issue is investigated.

**Comment by Wulfsta** [comment]
Using `flatbuffers` rather than `pkgs.flatbuffers` appears to fail. Reverting that change.

File: pkgs/development/python-modules/flatbuffers/default.nix (hunk 1 above)
Comment (Wulfsta) [reply pkgs/development/python-modules/flatbuffers/default.nix]: For this to work, we have to be explicit about what we're passing to the derivation in `pkgs/top-level/python-packages.nix`. This should be resolved in the latest commit.

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -2191,6 +2191,8 @@ in {
 
   flask_wtf = callPackage ../development/python-modules/flask-wtf { };
 
+  flatbuffers = callPackage ../development/python-modules/flatbuffers { inherit (pkgs) flatbuffers; };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:2194]: personal preference
```suggestion
  flatbuffers = callPackage ../development/python-modules/flatbuffers {
    inherit (pkgs) flatbuffers;
  };
```

File: pkgs/development/python-modules/flatbuffers/default.nix [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  pythonImportsCheck = [ "flatbuffers" ];
+
+  meta = {
+    description = "Python runtime library for use with the Flatbuffers serialization format";
+    homepage = flatbuf.meta.homepage;
+    license = flatbuf.meta.license;
+    maintainers = with stdenv.lib.maintainers; [ wulfsta ];
+  };
```
Comment (jonringer) [inline pkgs/development/python-modules/flatbuffers/default.nix:22]: personal preference
```suggestion
  meta = flatbuffers.meta // {
    description = "Python runtime library for use with the Flatbuffers serialization format";
    maintainers = with stdenv.lib.maintainers; [ wulfsta ];
  };
```

File: pkgs/development/python-modules/flatbuffers/default.nix [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+, flatbuffers
+}:
+
+let
+  flatbuf = flatbuffers;
+in buildPythonPackage rec {
+  pname = "flatbuffers";
+  version = flatbuf.version;
+
+  src = flatbuf.src;
```
Comment (jonringer) [inline pkgs/development/python-modules/flatbuffers/default.nix:12]: personal preference
```suggestion
buildPythonPackage rec {
  pname = "flatbuffers";
  inherit (flatbuffers) version src;

```

File: pkgs/top-level/python-packages.nix (hunk 4 above)
Comment (Wulfsta) [reply pkgs/top-level/python-packages.nix:2194]: Doing a cursory scroll through that file, it appears most imports like this are done in one line - I can change it if you like, but I do want to make sure first.

File: pkgs/development/python-modules/flatbuffers/default.nix (hunk 5 above)
Comment (Wulfsta) [reply pkgs/development/python-modules/flatbuffers/default.nix]: Should be resolved in the latest commit.

File: pkgs/development/python-modules/flatbuffers/default.nix (hunk 6 above)
Comment (Wulfsta) [reply pkgs/development/python-modules/flatbuffers/default.nix]: Should be resolved in the latest commit.

File: pkgs/top-level/python-packages.nix (hunk 4 above)
Comment (jonringer) [reply pkgs/top-level/python-packages.nix:2194]: If it's short enough, personally, I would like to keep under 80 characters when possible. Like I said, it's just my preference, but I'm not adamant about it.

Comment (Wulfsta) [reply pkgs/top-level/python-packages.nix:2194]: I'll change it then, not really a big deal either way to me. I think this derivation should be good to go after that.

**Review by jonringer — COMMENTED** [review]
tensorflow is now picking this up instead of the flatbuffers at the `pkgs` package scope. please apply this diff:
```
diff --git a/pkgs/top-level/python-packages.nix b/pkgs/top-level/python-packages.nix
index d9862a193e4..2432d3335c5 100644
--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -7063,6 +7063,7 @@ in {
     nccl = pkgs.nccl_cudatoolkit_10;
     openssl = pkgs.openssl_1_1;
     inherit (pkgs.darwin.apple_sdk.frameworks) Foundation Security;
+    inherit (pkgs) flatbuffers;
   };

   tensorflow-build_2 = callPackage ../development/python-modules/tensorflow/2 {
@@ -7072,6 +7073,7 @@ in {
     nccl = pkgs.nccl_cudatoolkit_11;
     openssl = pkgs.openssl_1_1;
     inherit (pkgs.darwin.apple_sdk.frameworks) Foundation Security;
+    inherit (pkgs) flatbuffers;
   };

   tensorflow-build = self.tensorflow-build_1;
```

**Review by jonringer — APPROVED** [review]
LGTM

```
https://github.com/NixOS/nixpkgs/pull/103554
3 packages built:
python27Packages.flatbuffers python37Packages.flatbuffers python38Packages.flatbuffers
```

**Pull request merged**